streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java

```diff
@@ -0,0 +1,245 @@
+    private Long timestamp;
+
+    public void setRecordMetadata(final String topic, final int partition, final long offset, final long timestamp) {
+        this.topic = topic;
+        this.partition = partition;
+        this.offset = offset;
+        this.timestamp = timestamp;
+    }
+    public void setTopic(final String topic) { this.topic = topic;}
+    public void setPartition(final int partition) { this.partition = partition;}
+    public void setOffset(final long offset) { this.offset = offset;}
+    public void setTimestamp(final long timestamp) { this.timestamp = timestamp;}
+
+    @Override public String topic() { return topic;}
+    @Override public int partition() { return partition;}
+    @Override public long offset() { return offset;}
+    @Override public long timestamp() { return timestamp;}
+
+    // mocks ================================================
+
+    // state stores
+    private final Map<String, StateStore> stateStores = new HashMap<>();
+    @Override
+    public void register(final StateStore store, final boolean loggingEnabledIsDeprecatedAndIgnored, final StateRestoreCallback stateRestoreCallbackIsIgnoredInMock) {
+        stateStores.put(store.name(), store);
+    }
+
+    @Override public StateStore getStateStore(final String name) { return stateStores.get(name);}
+
+    // punctuators
+    public static class CapturedPunctuator {
+        private final long intervalMs;
+        private final PunctuationType type;
+        private final Punctuator punctuator;
+
+        private CapturedPunctuator(final long intervalMs, final PunctuationType type, final Punctuator punctuator) {
+            this.intervalMs = intervalMs;
+            this.type = type;
+            this.punctuator = punctuator;
+        }
+
+        public long getIntervalMs() { return intervalMs;}
+        public PunctuationType getType() { return type;}
+        public Punctuator getPunctuator() { return punctuator;}
+    }
+
+    private final List<CapturedPunctuator> punctuators = new LinkedList<>();
+
+    @Override
+    public Cancellable schedule(final long intervalMs, final PunctuationType type, final Punctuator callback) {
+        punctuators.add(new CapturedPunctuator(intervalMs, type, callback));
+        return new Cancellable() {
+            @Override public void cancel() {}
+        };
+    }
+    @Override
+    public void schedule(final long interval) {
+        throw new UnsupportedOperationException("schedule() is deprecated and not supported in Mock. Use schedule(final long intervalMs, final PunctuationType type, final Punctuator callback) instead.");
+    }
+
+    public List<CapturedPunctuator> scheduledPunctuators() {
+        final LinkedList<CapturedPunctuator> capturedPunctuators = new LinkedList<>();
+        capturedPunctuators.addAll(punctuators);
+        return capturedPunctuators;
+    }
+
+    // forwards
+    private class Capture {
+        /*Nullable*/ private final Integer childIndex;
```

Comment:
Roger the timestamps.

I've added deprecation notices to these methods. I think as long as the corresponding methods are still on the Processor, we should let the mock capture them. 

ChildIndex and childName may be in the code base for a long time, for one thing. For another, I want migrating tests to the mock to be a pleasant experience. Plus, folks will already get compiler and IDE warnings about the deprecated methods.

The one exception is `public void schedule(final long interval)`, but it's because I just cannot support it in a mock. The fact that it's deprecated gives me an excuse to not support it.

Reply:
\cc @bbejeck @guozhangwang @dguy WDYT?

Reply:
For this specific API, I suspect it is ever commonly used in PAPI, so I'm fine with not supporting it right away, also as a way to encourage users to change code sooner than later, if there's anyone.

Reply:
I'm +1 on supporting the timestamps, even if it's not commonly used now, users will often look to tests for example usage (at least I do).  I'm also +1 on removing `childIndex` for the same reason, but I don't have too strong an opinion on that.